src/backend/server.js

```diff
@@ -131,10 +142,61 @@ function createServer({ privateStorage, sharedStorage, flavor }) {
       rnBridge.channel.on("sync-join", joinSync);
       rnBridge.channel.on("sync-leave", leaveSync);
       rnBridge.channel.on("replace-config", replaceConfig);
+      log("++++ 1 backend listening");
+      rnBridge.channel.once("apk-ready", apkReady);
+      rnBridge.channel.post("copy-apk");
       origListen.apply(server, args);
     });
   };
 
+  function apkReady(apkPath, version) {
+    log("++++ 2", apkPath, version);
+    if (!upgradeManager) {
+      createUpgradeManager(version, err => {
+        if (err) return onError("createUpgradeManager", err);
+        log("++++ 3 created upgrade manager");
+        upgradeManager.setApkInfo(apkPath, version, err => {
+          if (err) return onError("setApkInfo", err);
+          log("++++ 4 set apk info");
+          rnBridge.channel.on("p2p-upgrades-frontend-ready", () => {
+            rnBridge.channel.post("p2p-upgrades-backend-ready");
+            // Now we know the frontend is definitely ready!
+            log("++++ 5 frontend told us they are ready");
+          });
+          rnBridge.channel.post("p2p-upgrades-backend-ready");
+        });
+      });
+    }
+  }
+
+  function onError(prefix, err) {
+    if (!err) return;
+    main.bugsnag.notify(err, {
+      severity: "error",
+      context: prefix,
+    });
+    log(`error(${prefix}): ' + ${err.toString()}`);
+  }
+
+  function createUpgradeManager(version, cb) {
+    getport((err, port) => {
+      if (err) return cb(err);
+
+      const emitFn = rnBridge.channel.post.bind(rnBridge.channel);
+      const listenFn = rnBridge.channel.on.bind(rnBridge.channel);
+      const removeFn = rnBridge.channel.removeListener.bind(rnBridge.channel);
+      upgradeManager = new UpgradeManager(
+        upgradePath,
+        port,
+        version,
+        emitFn,
+        listenFn,
+        removeFn
+      );
```

Comment:
This architecture makes me nervous... I came here looking for the boundary between UpgradeManager and the `backend/server.js` code that hooks up the front-end. Giving UpgradeManager responsibility for managing listeners on the IPC channel feels prone to bugs, and figuring out exactly when and how events are emitted to the front-end is harder to follow. Is there a different way of implementing this?